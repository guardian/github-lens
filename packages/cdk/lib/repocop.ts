import { GuScheduledLambda, type GuScheduledLambdaProps } from '@guardian/cdk';
import type {
	GuLambdaErrorPercentageMonitoringProps,
	NoMonitoring,
} from '@guardian/cdk/lib/constructs/cloudwatch';
import type { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import { Duration } from 'aws-cdk-lib';
import type { IVpc, SecurityGroup } from 'aws-cdk-lib/aws-ec2';
import type { Schedule } from 'aws-cdk-lib/aws-events';
import { PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import type { DatabaseInstance } from 'aws-cdk-lib/aws-rds';
import { Secret } from 'aws-cdk-lib/aws-secretsmanager';
import { Topic } from 'aws-cdk-lib/aws-sns';
import type { ITopic } from 'aws-cdk-lib/aws-sns';
import { LambdaSubscription } from 'aws-cdk-lib/aws-sns-subscriptions';

export class Repocop {
	constructor(
		guStack: GuStack,
		schedule: Schedule,
		anghammaradTopic: ITopic,
		cloudqueryDB: DatabaseInstance,
		monitoringConfiguration:
			| NoMonitoring
			| GuLambdaErrorPercentageMonitoringProps,
		vpc: IVpc,
		interactiveMonitorTopic: Topic,
		dbSecurityGroup: SecurityGroup,
		repocopGithubSecret: Secret,
	) {
		const snykIntegratorInputTopic = new Topic(
			guStack,
			`snyk-integrator-input-topic-${guStack.stage}`,
			{
				displayName: 'Snyk Integrator Input Topic',
			},
		);

		const repocopLampdaProps: GuScheduledLambdaProps = {
			app: 'repocop',
			fileName: 'repocop.zip',
			handler: 'index.main',
			memorySize: 1024,
			monitoringConfiguration,
			rules: [{ schedule }],
			runtime: Runtime.NODEJS_20_X,
			environment: {
				ANGHAMMARAD_SNS_ARN: anghammaradTopic.topicArn,
				DATABASE_HOSTNAME: cloudqueryDB.dbInstanceEndpointAddress,
				QUERY_LOGGING: 'false', // Set this to 'true' to enable SQL query logging
				INTERACTIVE_MONITOR_TOPIC_ARN: interactiveMonitorTopic.topicArn,
				GITHUB_APP_SECRET: repocopGithubSecret.secretArn,
				INTERACTIVES_COUNT: guStack.stage === 'PROD' ? '40' : '3',
			},
			vpc,
			securityGroups: [dbSecurityGroup],
			timeout: Duration.minutes(5),
		};

		const repocopLambda = new GuScheduledLambda(
			guStack,
			'repocop',
			repocopLampdaProps,
		);

		//allow repocop to create and put metrics
		const policyStatement = new PolicyStatement({
			actions: ['cloudwatch:PutMetricData'],
			resources: ['*'],
		});

		cloudqueryDB.grantConnect(repocopLambda, 'repocop');

		repocopGithubSecret.grantRead(repocopLambda);
		anghammaradTopic.grantPublish(repocopLambda);
		interactiveMonitorTopic.grantPublish(repocopLambda);
		repocopLambda.addToRolePolicy(policyStatement);

		const snykIntegratorSecret = new Secret(
			guStack,
			`snyk-integrator-github-app-auth`,
			{
				secretName: `/${guStack.stage}/${guStack.stack}/service-catalogue/snyk-integrator-github-app-secret`,
			},
		);

		const snykIntegatorLambda: GuLambdaFunction = new GuLambdaFunction(
			guStack,
			'snyk-integrator',
			{
				app: 'snyk-integrator',
				fileName: 'snyk-integrator.zip',
				handler: 'index.handler',
				memorySize: 1024,
				runtime: Runtime.NODEJS_20_X,
				environment: {
					GITHUB_APP_SECRET: snykIntegratorSecret.secretArn,
				},
				vpc,
				timeout: Duration.minutes(5),
			},
		);

		snykIntegratorSecret.grantRead(snykIntegatorLambda);
		snykIntegratorInputTopic.addSubscription(
			new LambdaSubscription(snykIntegatorLambda, {}),
		);
	}
}
