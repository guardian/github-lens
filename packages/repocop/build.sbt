ThisBuild / version := "0.1.0-SNAPSHOT"

ThisBuild / scalaVersion := "3.2.1"

lazy val root = (project in file("."))
  .settings(
    name := "repocop",
    idePackagePrefix := Some("com.gu.repocop")
  )
val circeVersion = "0.14.3"

libraryDependencies ++= Seq(
  "io.circe" %% "circe-core" % circeVersion,
  "io.circe" %% "circe-generic" % circeVersion,
  "io.circe" %% "circe-parser" % circeVersion,
  "com.lihaoyi" %% "requests" % "0.7.1",
  "org.scalatest" %% "scalatest" % "3.2.14" % Test
)

val jarName = "repocop.jar"
assembly / assemblyJarName := jarName
