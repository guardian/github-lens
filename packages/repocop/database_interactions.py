import boto3
import json
import pandas as pd
import psycopg2 as ps
import os

def _get_secret_arn(secret_client) -> str:
    paginator = secret_client.get_paginator('list_secrets')

    response_iterator = paginator.paginate(
        Filters=[
            {
                'Key': 'tag-value',
                'Values': [
                    'cloud-query' #App
                ]
            }
        ]
    )
    secret_arn = list(response_iterator)[0]["SecretList"][0]['ARN']
    return secret_arn


def connect_to_db(session: boto3.Session):
    """Gets the credentials for the database from AWS Secrets Manager"""
    secret_client = session.client('secretsmanager', 'eu-west-1')

    if "SECRET_ARN" in os.environ:
        print('AWS Lambda environment detected')
        secret_arn = os.environ.get('SECRET_ARN')
    else:
        print("Local environment detected: fetching secret ARN")
        secret_arn = _get_secret_arn(secret_client)

    secret = secret_client.get_secret_value(
        SecretId=secret_arn
    )["SecretString"]

    secret_json = json.loads(secret)
    conn = ps.connect(
    host=secret_json['host'],
    port=secret_json['port'],
    dbname='',
    user=secret_json['username'],
    password=secret_json['password'])

    return conn

def select(table_name:str, columns: list[str], db_engine) -> pd.DataFrame:
    return pd.read_sql_table(table_name, con=db_engine, columns=columns)