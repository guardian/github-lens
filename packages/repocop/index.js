"use strict";
(() => {
  // repocop.ts
  console.log("Hello repocop");
})();
