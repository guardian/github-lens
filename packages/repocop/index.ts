import { PrismaClient } from "@prisma/client";
import {Signer} from "@aws-sdk/rds-signer";

interface Config {
	stage: string;
	database: {
		hostname: string;
		user: string;
		port: number;
	}
}

function getEnvOrThrow(key: string): string {
	const value = process.env[key];
	if (value === undefined) {
		throw new Error(`Environment variable ${key} is not set`);
	}
	return value;
}
function getConfig(): Config {
	return {
		stage: process.env['STAGE'] || 'DEV',
		database: {
			hostname: getEnvOrThrow('DATABASE_HOSTNAME'),
			user: 'repocop',
			port: 5432,
		}
	}
}


export async function main() {
	const config = getConfig();
	const prisma = new PrismaClient({
		datasources: {
			db: {
				url: await getDatabaseUrl(config),
			},
		}
	});
	console.log('Query prisma');
	const queryResult = await prisma.github_repositories.findFirst({where: {
		archived: false,
	}})

	if (queryResult)
		console.log(queryResult.name || "not found")
}

async function getDatabaseUrl(config: Config){

	const {hostname, port, user} = config.database

	const signer = new Signer({
		hostname,
		port,
		username: user,
		region: 'eu-west-1'
	});
	const token = await signer.getAuthToken();

	const databaseUrl = `postgres://repocop:${encodeURIComponent(token)}@${hostname}:${port}/postgres?schema=public&sslmode=verify-full`;

	return databaseUrl
}


