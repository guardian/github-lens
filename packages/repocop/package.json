{
	"name": "repocop",
	"version": "1.0.0",
	"devDependencies": {
		"@types/aws-lambda": "^8.10.122",
		"prisma": "^5.4.2"
	},
	"scripts": {
		"build": "esbuild src/index.ts --bundle --platform=node --target=node18 --outdir=dist --external:@aws-sdk --external:@prisma/client --external:prisma",
		"start": "ts-node src/run-locally.ts",
		"postinstall": "prisma generate",
		"test": "jest --detectOpenHandles --config ../../jest.config.js --selectProjects repocop",
		"migrate:setuplocal": "script/migrate.sh --dev",
		"migrate:dev": "script/migrate.sh --dev $NAME",
		"migrate:code": "script/migrate.sh --code",
		"migrate:prod": "script/migrate.sh --prod"
	},
	"prisma": {
		"schema": "prisma/schema.prisma"
	},
	"dependencies": {
		"@aws-sdk/client-sqs": "^3.423.0",
		"@aws-sdk/rds-signer": "^3.429.0",
		"@prisma/client": "^5.4.2"
	}
}
