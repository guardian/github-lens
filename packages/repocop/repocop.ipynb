{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Repocop prototyping notebook\n",
    "\n",
    "> **Warning**\n",
    "> This notebook writes to the cloudquery database. Proceed with extreme caution\n",
    "\n",
    "> **Warning**\n",
    "> Please clear all your outputs and restart the kernel before committing\n",
    "\n",
    "### Prerequisites\n",
    "- Have deployTools credentials\n",
    "- Be connected to the VPN\n",
    "\n",
    "### Summary\n",
    "This notebook connects to the cloudquery database, evaluates resources (aka github repositories) against best practice rules for those resources, and writes the information as a table back to the db, so other tools can use the information to notify teams about actions they need to take."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlalchemy as sa\n",
    "import boto3\n",
    "from database_interactions import connect_to_db\n",
    "\n",
    "#create the database connection\n",
    "session = boto3.Session(profile_name='deployTools')\n",
    "conn = connect_to_db(session)\n",
    "engine = sa.create_engine('postgresql://', creator=lambda: conn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we create our first intermediate dataframe, assessing rule REPOSITORY-01, which says that the default branch for all repositories should be main.\n",
    "\n",
    "n.b. we have not yet exempted archived repositories from any of these rulesets in code. A decision about where to deal with those has yet to be made"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from database_interactions import select\n",
    "import pandas as pd\n",
    "\n",
    "#default_branch_df needs to be a dataframe with the following columns:\n",
    "# - full_name (string)\n",
    "# - default_branch (string)\n",
    "def repository_01(default_branch_df: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"Repository 01: Default branch should be main\"\"\"\n",
    "    default_branch_df['repository_01'] = default_branch_df['default_branch'] == 'main'\n",
    "    return default_branch_df[['full_name', 'repository_01']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "default_branch_df = select('github_repositories',['full_name', 'default_branch'], engine) #this includes archived repos. make a decision about this\n",
    "repository_01_df = repository_01(default_branch_df)\n",
    "repository_01_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we asses rule REPOSITORY_06, which is significantly more complicated. It states that all repositories owned by one or more P+E teams require a valid production status topic. Repos owned exlusively by teams outside of P+E are exempt from this requirement. The most common example of this is interactives repositories.\n",
    "\n",
    "Production statuses recognised by DevX are enumerated in the `guardian_production_status` table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from rules import repository_06\n",
    "\n",
    "#gather all the required data for repository_06\n",
    "valid_topics_df = select('guardian_production_status', ['status', 'priority'], engine)\n",
    "topics_list = valid_topics_df['status'].tolist()\n",
    "non_pe_teams_list = select('guardian_non_p_and_e_github_teams', ['team_name'], engine)['team_name'].tolist()\n",
    "topics_df = select('github_repositories', ['full_name', 'topics'], engine)\n",
    "teams_df = select('github_teams', ['name', 'slug'], engine)\n",
    "#select function doesn't work on views, so we have to use read_sql_query\n",
    "ownership_df = pd.read_sql_query(\"select repo_name, github_team_name, github_team_id from view_repo_ownership\", con=conn)\n",
    "ownership_df = ownership_df.merge(teams_df, how='left', left_on='github_team_name', right_on='name')[['repo_name', 'github_team_name', 'slug']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "freq = ownership_df['github_team_name'].value_counts()\n",
    "freq[:15].plot(kind='bar', title='Teams that own the most repos', xlabel='team name', ylabel='Count')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "repository_06_df = repository_06(ownership_df, topics_df, topics_list, non_pe_teams_list)\n",
    "freq = repository_06_df['repository_06'].value_counts()\n",
    "freq.plot.pie(subplots=True, figsize=(5, 4), title='repository_06')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we merge the interim rule tables into one table, keyed on repository name, and write that information back to the database. \n",
    "\n",
    "### Please exercise extreme caution before modifying this section as we have not set up a local development environment yet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "repository_rule_df =  repository_01_df.merge(repository_06_df, how='left', left_on='full_name', right_on='repo_name')[['full_name', 'repository_01', 'repository_06']]\n",
    "repository_rule_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "repository_rule_df.to_sql('repocop_github_repository_rules', engine, if_exists='replace', index=False)\n",
    "pd.read_sql_query(\"select * from repocop_github_repository_rules\", con=conn).head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
