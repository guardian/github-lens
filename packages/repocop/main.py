import boto3
from database_interactions import connect_to_db, select
import datetime as dt
from rules import repository_01
import sqlalchemy as sa

def handler(event=None, context=None):
    session = boto3.Session(region_name='eu-west-1')
    conn = connect_to_db(session)
    engine = sa.create_engine('postgresql://', creator=lambda: conn)
    print("Connected to CloudQuery DB")
    default_branch_df = select('github_repositories',['full_name', 'default_branch'], engine)
    repository_01_df = repository_01(default_branch_df)
    repository_01_df['timestamp'] = dt.datetime.now().replace(microsecond=0)
    repository_01_df.to_sql('repocop_github_repository_rules', engine, if_exists='replace', index=False)
    print("Successfully updated repocop_github_repository_rules table")

if __name__ == "__main__":
    handler()
