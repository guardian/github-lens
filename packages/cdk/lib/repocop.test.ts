import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Repocop } from './repocop';

describe('The Repocop stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new Repocop(app, 'GithubLens', {
			stack: 'deploy',
			stage: 'INFRA',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
