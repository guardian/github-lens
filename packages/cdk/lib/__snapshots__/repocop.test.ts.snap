// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Repocop stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuVpcParameter",
      "GuSubnetListParameter",
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
      "GuLambdaErrorPercentageAlarm",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "VpcId": {
      "Default": "/account/vpc/primary/id",
      "Description": "Virtual Private Cloud to run EC2 instances within. Should NOT be the account default VPC.",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::VPC::Id>",
    },
    "repocopPrivateSubnets": {
      "Default": "/account/vpc/primary/subnets/private",
      "Description": "A list of private subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
  },
  "Resources": {
    "repocoplambdaB702A335": {
      "DependsOn": [
        "repocoplambdaServiceRoleDefaultPolicyCD286095",
        "repocoplambdaServiceRole65643A09",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "deploy/INFRA/repocop/repocop.jar",
        },
        "Environment": {
          "Variables": {
            "APP": "repocop",
            "STACK": "deploy",
            "STAGE": "INFRA",
          },
        },
        "Handler": "com.gu.repocop.Handler::handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "repocoplambdaServiceRole65643A09",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Tags": [
          {
            "Key": "App",
            "Value": "repocop",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/service-catalogue",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "INFRA",
          },
        ],
        "Timeout": 300,
        "VpcConfig": {
          "SecurityGroupIds": [
            {
              "Fn::GetAtt": [
                "repocoplambdaSecurityGroupF34437C3",
                "GroupId",
              ],
            },
          ],
          "SubnetIds": {
            "Ref": "repocopPrivateSubnets",
          },
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "repocoplambdaErrorPercentageAlarmForLambda7BA16892": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":devx-alerts",
              ],
            ],
          },
        ],
        "AlarmDescription": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "repocoplambdaB702A335",
              },
              " exceeded 0% error rate",
            ],
          ],
        },
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              "High error % from ",
              {
                "Ref": "repocoplambdaB702A335",
              },
              " lambda in INFRA",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "100*m1/m2",
            "Id": "expr_1",
            "Label": {
              "Fn::Join": [
                "",
                [
                  "Error % of ",
                  {
                    "Ref": "repocoplambdaB702A335",
                  },
                ],
              ],
            },
          },
          {
            "Id": "m1",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "FunctionName",
                    "Value": {
                      "Ref": "repocoplambdaB702A335",
                    },
                  },
                ],
                "MetricName": "Errors",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "m2",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "FunctionName",
                    "Value": {
                      "Ref": "repocoplambdaB702A335",
                    },
                  },
                ],
                "MetricName": "Invocations",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "repocoplambdaSecurityGroupF34437C3": {
      "Properties": {
        "GroupDescription": "Automatic security group for Lambda Function RepocoprepocoplambdaD20097B3",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "repocop",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/service-catalogue",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "INFRA",
          },
        ],
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "repocoplambdaServiceRole65643A09": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "repocop",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/service-catalogue",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "INFRA",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "repocoplambdaServiceRoleDefaultPolicyCD286095": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/deploy/INFRA/repocop/repocop.jar",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/INFRA/deploy/repocop",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/INFRA/deploy/repocop/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "repocoplambdaServiceRoleDefaultPolicyCD286095",
        "Roles": [
          {
            "Ref": "repocoplambdaServiceRole65643A09",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "repocoplambdarepocoplambdacron080759EAEC6": {
      "Properties": {
        "ScheduleExpression": "cron(0 8 * * ? *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "repocoplambdaB702A335",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "repocoplambdarepocoplambdacron080AllowEventRuleRepocoprepocoplambdaD20097B3E7516549": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "repocoplambdaB702A335",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "repocoplambdarepocoplambdacron080759EAEC6",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
